### ISSUE TITLE
Contradictory statements on order of lookups

### ISSUE THREAD

**Issue by NorbertLindenberg** [MicrosoftDocs/typography-issues#380]
The document “OpenType Layout Common Table Formats” states: “To implement features, a client applies the lookups in the order the lookup definitions occur in the LookupList. As a result, within the GSUB or GPOS table, lookups from several different features may be interleaved during text processing.”

This statement contradicts statements in the descriptions of most shaping engines, which describe several sequential stages of feature application, each with a defined set of features to be applied within the stage, and sometimes requiring that features are applied one by one.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1cb42967-20ca-9270-4572-65315127fcc5
* Version Independent ID: 1fa413f1-71cf-ac2a-0999-70dd5d8ac4cd
* Content: [OpenType layout common table formats - Typography](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#slTbl_sRec)
* Content Source: [typographydocs/opentype/spec/chapter2.md](https://github.com/MicrosoftDocs/typography/blob/live/typographydocs/opentype/spec/chapter2.md)
* Product: **typography**

**Comment by PeterCon** [comment]
Not really contradictory within the OT spec, but does ignore the larger context that a client may process certain sets of features in sequential stages. That would be good clarification to add.

**Comment by NorbertLindenberg** [comment]
Is there any relevant client that “applies the lookups in the order the lookup definitions occur in the LookupList”?

**Comment by PeterConstable** [comment]
Every implementation should for whatever features get processed in the same shaping stage.

**Comment by NorbertLindenberg** [comment]
Let me rephrase my question using your terminology: Is there any relevant client that does not break the lookups in the LookupList into separate sets of features in sequential stages?

Also, is there any shaping engine that, for all stages that it defines, “applies the lookups in the order the lookup definitions occur in the LookupList”?

**Comment by PeterCon** [comment]
Let's take [feature application in USE](https://docs.microsoft.com/en-us/typography/script-development/use#opentype-feature-application-i) as an example: during GSUB processing, it processes features in six groups,

1. First, "Default glyph pre-processing group": locl, ccmp, nukt, akhn
2. Second, "Reordering group": rphf, pref
3. Third, "Orthographic unit shaping group"
4. Fourth (after a re-ordering pass), "Topographical features, GSUB": isol, init, medi, fina
5. Fifth, "Standard typographic presentation, GSUB": abvs, blws, calt, clig, haln, liga, pres, psts, rclt, rlig, vert, vrt2
6. Sixth, "Custom substitution features requested by the application"

(Actually, I would have thought groups 5 and 6 would be combined, since group 5 includes features like liga that are optional but on by default. But issues for the USE spec would be separate from issues for this part of the OT spec.)

Within each group, features get associated with some sub-runs within the string, and then lookup processing is done across the string. Whatever lookups are associated with the features in that group, all of those lookups will get processed one at a time in the order they occur within the LookupList table.

So, for example, suppose the Feature tables in the feature list for features in the first group have lookupListIndices[] arrays as follows:

* akhn: `lookupListIndices = [17, 22]`  
* ccmp: `lookupListIndices = [19, 8]`  
* locl: `lookupListIndices = [2, 31, 10]`  
* nukt: `lookupListIndices = [12, 18]`  

Then in that feature-processing stage, the following lookups would be applied in this order:

* 2, 8, 10, 12, 17, 18, 19, 22, 31

That's how all "clients" should be implemented.

What I was meaning earlier is that this is consistent with what is currently stated in the OT spec, but that the OT spec doesn't clarify that clients may partition features into different groups that each get processed in stages, one group of features at a time. So that's what I was thinking out to be clarified in the spec.

**Comment by PeterCon** [comment]
Addendum: the USE spec (in describing the first group) has a statement on the order of lookup processing that reflects what the OT spec says:

> Lookups associated with any of these features will be triggered in the lookup-order specified by the font developer. It is possible to interleave lookups for different features. 

Now, that is immediately followed by this:

>The order of features given here is recommended.

I don't understand what the intent of that is. The OT spec states, regarding the FeatureList,

>The FeatureRecord array is arranged alphabetically by FeatureTag names.

So that last statement in the USE spec doesn't seem to make sense to me. But that is an issue for the USE spec, not the OT spec.

**Comment by tiroj** [comment]
>_“To implement features, a client applies the lookups in the order the lookup definitions occur in the LookupList. As a result, within the GSUB or GPOS table, lookups from several different features may be interleaved during text processing.”

This text is interpretable in a way that makes it technically accurate, but it is imprecise. It should convey that some features are applied by shaping engines in set order or in processing blocks, but that within processing blocks LookupList order determines the order in which the client applies lookups and, hence, within processing blocks lookups from different features may be interleaved.

**Comment by tiroj** [comment]
> So that last statement in the USE spec doesn't seem to make sense to me. But that is an issue for the USE spec, not the OT spec.

It’s shorthand for ‘the order of lookups associated to features given here is recommended’. It’s a perennial problem in talking about OTL that features are the thing with shared labels between fonts—while the lookups that are actually what we often need to talk about have either arbitrary names in sources or numbers in the LookupList. So we fall back on talking about feature order when what we're really talking about is the order of the lookups associated to the features.

**Comment by NorbertLindenberg** [comment]
> Whatever lookups are associated with the features in that group, all of those lookups will get processed one at a time in the order they occur within the LookupList table.

Not true for the reordering group of the USE – the documentation says “These features are applied individually in this order, rphf, pref”.

**Comment by PeterCon** [comment]
>This text is interpretable in a way that makes it technically accurate, but it is imprecise...

As the quoted text was from the USE spec, please open an issue on that as you deem appropriate. (No point discussing further here.)

**Comment by PeterCon** [comment]
@NorbertLindenberg 
>Not true for the reordering group of the USE...

I amended my earlier comment to address that case. Please see the revised comment. My general point still holds.

**Comment by NorbertLindenberg** [comment]
> I amended my earlier comment to address that case. Please see the revised comment.

Got it – sorry for missing that earlier.

> My general point still holds.

You’re still saying that clients “may partition features”. We haven’t identified any client that does not partition them, so the spec should be clear that partitioning is the normal case, with a remote possibility that some client may lump all features into the same subset.

**Comment by PeterCon** [comment]
I think "may" is sufficient, but the wording could indicate it's not _un_common. IIRC, MS's "generic" engine which was used for Latin, Cyrillic, and other scripts did not do any partitioning of features in that way. So, I'm sure it's not true for all "clients".

In any case, as long as this spec points out that that's a permitted possibility, it's up to shaping specs to indicate what partitioning if any is done.

**Comment by tiroj** [comment]
I think even the most generic shaping engines might still be processing locl and ccmp before other features. I think it would be worthwhile to point out that shaping engines commonly prioritise some features independently of lookup order, and that shaping for some scripts requires processing blocks of features (e.g. pre- and post-reordering blocks for Indic scripts).

**Comment by PeterCon** [comment]
>I think it would be worthwhile to point out that shaping engines commonly prioritise some features independently of lookup order, and that shaping for some scripts requires processing blocks of features (e.g. pre- and post-reordering blocks for Indic scripts).

That's along the lines of what I was thinking would be appropriate.

**Comment by PeterCon** [comment]
Proposed change for the next version: add the following note in the *Features and Lookups* section after the note about GSUB preceding GPOS:

>><ins>*Note:* A client may process GSUB features in ordered stages, with particular features processed in each stage. A client may also perform certain operations on glyph sequences, such as re-ordering of glyphs, before or between these stages. Such processing is required for correct support of many scripts. Details on such script-specific processing is outside the scope of this specification. Within any such feature-processing stage, however, lookups referenced by the features applied in that stage must be processed in LookupList order.</ins>

**Comment by behdad** [comment]
The "may" wording is problematic because it gives the impression that it's up to the client to choose what to do.

I suggest a more direct note saying that "many script-specific shaping engines process features in stages and only within each stage the lookups are applied by increasing lookup-index."

**Comment by PeterCon** [comment]
I don't think the "may" wording is problematic here since this is a very general context—what is true of feature/lookup processing in the general case. It's noted that this will be required for implementation of some scripts, and leaves the details to some separate specification(s).

**Comment by NorbertLindenberg** [comment]
There seems to be an implicit assumption here that shaping engines cannot be mentioned in the “spec” part of the OpenType documentation. Why is that? It should be clear to everybody that you can’t have a functional OpenType implementation without shaping engines (and assorted shaping-related functionality that’s yet to be  documented), and that you can’t create a font for any but the simplest scripts if you don’t understand the shaping engine(s) supporting the script(s) of the font.

**Comment by PeterCon** [comment]
I don't see how it will make a lick of difference here if "shaping engines" are mentioned instead of "applications". But it would introduce a need in this spec to explain what is meant by the term. I don't see the point.

**Comment by NorbertLindenberg** [comment]
Applications using OpenType number in the millions, and we don’t know their developers, so it’s impossible to know what they all do and impossible to standardize their behavior. As a result, we’re stuck with “may” wording and fuzzy recommendations.

Rendering systems with shaping engines number in the single digits (I know of five), and we mostly know their developers, so it’s quite possible to find out what they do and to standardize them. It allows us to describe reality, as in “Most clients process GSUB features in ordered stages...”.

There are many other parts of the spec that would benefit from this improvement in clarity.

**Comment by PeterCon** [comment]
I'm not opposed to the direction you're suggesting... at some point. But this change in this one place would have no effect, and would just make the spec _less_ coherent. To have any significant value it would need to be something done systematically. That means touching probably hundreds of places in the spec, as well as a coherent discussion somewhere of the architectural model that's assumed. Even if the change weren't that big, the ROI is unclear: it's not going to change the handful of shaping engine implementations, or any font development tools.

So, sure, at some point. I'm not inclined to start piecemeal now. I'd give much higher priority to dealing with issues that really are impacting implementations now; e.g., #191, #310, #363, #407, #455, #535, #560, #569, #576, and several others. Many of these require time to get input and discussion, so I can work on some lower impact suggestions in the meantime; like #347, #509, maybe even #519 (a more localized but still very big change that you've also asked for).
